### PR TITLE
feat(VoiceState): add methods for fetching voice state

### PR DESCRIPTION
**Pull request by imnaiyar**
**Please describe the changes this PR makes and why it should be merged:**
- Added methods for fetching VoiceState (see https://github.com/discord/discord-api-docs/pull/7061).
- Deprecated and moved `editUserVoiceState` and `setVoiceState` from GuildsAPI to VoiceAPI in core to be more in line with API documentations

Depends on
- https://github.com/discordjs/discord-api-types/pull/1046
- https://github.com/discord/discord-api-docs/pull/7061
- #10452 

**Status and versioning classification:**
 - Code changes have been tested against the Discord API, or there are no code changes
 - I know how to update typings and have done so, or typings don't need updating
 - This PR changes the library's interface (methods or parameters added)

### CODE REVIEW

**Review by sdanialraza — COMMENTED** [review]
`GatewayVoiceState` is now unused and can be removed from [here](https://github.com/imnaiyar/discord.js/blob/feat/voice-state/packages/discord.js/typings/rawDataTypes.d.ts#L66)

File: packages/discord.js/src/structures/VoiceState.js [hunk 1]
```diff
@@ -250,6 +250,15 @@ class VoiceState extends Base {
     return this;
   }
 
+  /**
+   * Fetches this voice state.
+   * @param {boolean} [force=true] Whether to skip the cache check and request the API
+   * @returns {Promise<VoiceState>}
+   */
+  fetch(force = true) {
+    return this.guild.voiceStates.fetch(this.member, { force });
```
Comment (Syjalo) [inline packages/discord.js/src/structures/VoiceState.js:259]: Yes, it resolves, but I think id would be more suitable
```suggestion
    return this.guild.voiceStates.fetch(this.id, { force });
```

**Review by Syjalo — CHANGES_REQUESTED** [review]

File: packages/discord.js/typings/index.d.ts [hunk 2]
```diff
@@ -3589,6 +3589,7 @@ export class VoiceState extends Base {
   public setRequestToSpeak(request?: boolean): Promise<this>;
   public setSuppressed(suppressed?: boolean): Promise<this>;
   public edit(options: VoiceStateEditOptions): Promise<this>;
+  public fetch(force?: boolean): Promise<this>;
```
Comment (Syjalo) [inline packages/discord.js/typings/index.d.ts:3592]: It doesn't return `this`
```suggestion
  public fetch(force?: boolean): Promise<VoiceState>;
```

File: packages/discord.js/src/managers/VoiceStateManager.js [hunk 3]
```diff
@@ -32,6 +33,27 @@ class VoiceStateManager extends CachedManager {
...
   }
+
+  /**
+   * Obtains a user's voice state from discord or from the cache if it's already available.
+   * @param {GuildMemberResolvable} member The member whose voice state is to be fetched
+   * @param {BaseFetchOptions} [options] Additional options for this fetch
+   * @returns {Promise<VoiceState>}
+   * @example
+   * // Fetch a member's voice state
+   * guild.voiceStates.fetch("851588007697580033")
```
Comment (Syjalo) [inline packages/discord.js/src/managers/VoiceStateManager.js:44]: The docs uses hydrabolt's id
```suggestion
   * guild.voiceStates.fetch("66564597481480192")
```

File: packages/discord.js/src/managers/VoiceStateManager.js [hunk 4]
```diff
@@ -32,6 +33,27 @@ class VoiceStateManager extends CachedManager {
...
+   *    .then(state => console.log(`Member's voice channel: ${state.channel}`))
```
Comment (Syjalo) [inline packages/discord.js/src/managers/VoiceStateManager.js:45]: Just log it
```suggestion
   *    .then(console.log)
```

File: packages/core/src/api/voice.ts [hunk 5]
```diff
@@ -15,4 +24,59 @@ export class VoiceAPI {
...
+	 *
+	 * @see {@link https://discord.com/developers/docs/resources/voice#modify-current-user-voice-state}
+	 * @param guildId - The id of the guild
+	 * @param body - The options for setting the voice state
+	 */
+	public async setVoiceState(guildId: Snowflake, body: RESTPatchAPIGuildVoiceStateCurrentMemberJSONBody = {}) {
+		return this.rest.patch(Routes.guildVoiceState(guildId, '@me'), {
+			body,
+		}) as Promise<RESTPatchAPIGuildVoiceStateCurrentMemberResult>;
+	}
```
Comment (Syjalo) [inline packages/core/src/api/voice.ts:81]: Shouldn't `PUT` method use `set`?
```suggestion
	/**
	 * Edits the voice state for the current user
	 *
	 * @see {@link https://discord.com/developers/docs/resources/voice#modify-current-user-voice-state}
	 * @param guildId - The id of the guild
	 * @param body - The options for editing the voice state
	 */
	public async editVoiceState(guildId: Snowflake, body: RESTPatchAPIGuildVoiceStateCurrentMemberJSONBody = {}) {
		return this.rest.patch(Routes.guildVoiceState(guildId, '@me'), {
			body,
		}) as Promise<RESTPatchAPIGuildVoiceStateCurrentMemberResult>;
	}
```

File: packages/core/src/api/guild.ts [hunk 6]
```diff
@@ -1298,9 +1300,10 @@ export class GuildsAPI {
 	/**
 	 * Sets the voice state for the current user
 	 *
-	 * @see {@link https://discord.com/developers/docs/resources/guild#modify-current-user-voice-state}
+	 * @see {@link https://discord.com/developers/docs/resources/voice#modify-current-user-voice-state}
 	 * @param guildId - The id of the guild
 	 * @param body - The options for setting the voice state
+	 * @deprecated Use {@link VoiceAPI.setVoiceState}
```
Comment (Syjalo) [inline packages/core/src/api/guild.ts:1306]:
```suggestion
	 * @deprecated Use {@link VoiceAPI.editVoiceState}
```

**Review by Syjalo — CHANGES_REQUESTED** [review]

File: packages/core/src/api/voice.ts (hunk 5 above)
Comment (imnaiyar) [reply packages/core/src/api/voice.ts:81]: This is how it was in `GuildsAPI`, I just deprecated it and moved it here, I'm not sure if we want to rename it.

**Comment by jaw0r3k** [comment]
You forgot to handle `@me` as a user parameter [(docs reference)](https://discord.com/developers/docs/resources/voice#get-current-user-voice-state)

File: packages/core/src/api/guild.ts [hunk 7]
```diff
@@ -687,11 +688,12 @@ export class GuildsAPI {
 	/**
 	 * Edits a user's voice state in a guild
 	 *
-	 * @see {@link https://discord.com/developers/docs/resources/guild#modify-user-voice-state}
+	 * @see {@link https://discord.com/developers/docs/resources/user#modify-user-voice-state}
```
Comment (Qjuh) [inline packages/core/src/api/guild.ts:691]:
```suggestion
	 * @see {@link https://discord.com/developers/docs/resources/voice#modify-user-voice-state}
```

**Review by Qjuh — CHANGES_REQUESTED** [review]

**Review by Qjuh — APPROVED** [review]

**Review by kyranet — APPROVED** [review]

**Review by SpaceEEC — APPROVED** [review]

**Review by almeidx — APPROVED** [review]

File: packages/core/src/api/guild.ts [hunk 8]
```diff
@@ -102,6 +102,7 @@ import {
 	type RESTPutAPIGuildTemplateSyncResult,
 	type Snowflake,
 } from 'discord-api-types/v10';
+import { VoiceAPI } from './voice';
```
[the remark this answers is not in the document]
Comment (vladfrangu) [reply packages/core/src/api/guild.ts:106]: Alternatively, shouldn't we redirect the function calls to the voice api ones?

[the remark this answers is not in the document]
Comment (imnaiyar) [reply packages/core/src/api/guild.ts:106]: > Alternatively, shouldn't we redirect the function calls to the voice api ones?

So something like
```ts
new VoiceAPI().editUserVoiceState(...)
```
?

[the remark this answers is not in the document]
Comment (vladfrangu) [reply packages/core/src/api/guild.ts:106]: Sure 🫡

**Review by vladfrangu — APPROVED** [review]

**Pull request merged**